### PR TITLE
ICMSLST-486 Edit Goods when preparing the licence (ILB Admin)

### PR DESCRIPTION
**Pull request by kevinetienne**
Edit goods for sanctions, derogations and woods. The goods can be edited in the response preparation but not the commodity code.

### CODE REVIEW

File: web/domains/case/_import/derogations/views.py [hunk 1]
```diff
@@ -236,5 +237,37 @@ def manage_checklist(request, pk):
         )
 
 
+@login_required
+@permission_required("web.reference_data_access", raise_exception=True)
+def edit_goods_licence(request: HttpRequest, pk: int) -> HttpResponse:
+    application: DerogationsApplication = get_object_or_404(DerogationsApplication, pk=pk)
+    with transaction.atomic():
+        task: Task = application.get_task(
```
Comment (oskusalerma) [inline web/domains/case/_import/derogations/views.py:245]: i don't think you need to type Task, application is typed, get_task is typed, mypy knows the type of task.

please use reveal_type(task) to verify i'm correct about this

File: web/domains/case/_import/sanctions/views.py [hunk 2]
```diff
@@ -116,42 +118,76 @@ def add_goods(request, pk):
...
+
+@login_required
+@permission_required("web.reference_data_access", raise_exception=True)
+def edit_goods_licence(request, application_pk, goods_pk):
+    application: SanctionsAndAdhocApplication = get_object_or_404(
+        SanctionsAndAdhocApplication, pk=application_pk
+    )
+
+    with transaction.atomic():
+        task: Task = application.get_task(
```
Comment (oskusalerma) [inline web/domains/case/_import/sanctions/views.py:146]: same

File: web/domains/case/_import/views.py [hunk 3]
```diff
@@ -632,10 +632,12 @@ def respond_fir(request, application_pk, fir_pk):
...
 @permission_required("web.reference_data_access", raise_exception=True)
-def prepare_response(request, pk):
+def prepare_response(request: HttpRequest, pk: int) -> HttpResponse:
     with transaction.atomic():
-        application = get_object_or_404(ImportApplication.objects.select_for_update(), pk=pk)
-        task = application.get_task(
+        application: ImportApplication = get_object_or_404(
+            ImportApplication.objects.select_for_update(), pk=pk
+        )
+        task: Task = application.get_task(
```
Comment (oskusalerma) [inline web/domains/case/_import/views.py:640]: same

File: web/domains/case/_import/views.py [hunk 4]
```diff
@@ -649,21 +651,74 @@ def prepare_response(request, pk):
...
+    elif application.process_type == DerogationsApplication.PROCESS_TYPE:
+        return _prepare_derogations_response(request, application.derogationsapplication, context)
+    elif application.process_type == WoodQuotaApplication.PROCESS_TYPE:
+        return _prepare_wood_quota_response(request, application.woodquotaapplication, context)
+    else:
+        raise NotImplementedError
+
+
+def _prepare_response_oil(
+    request: HttpRequest, application: OpenIndividualLicenceApplication, context: dict
```
Comment (oskusalerma) [inline web/domains/case/_import/views.py:675]: context could be Dict[str, Any]

File: web/domains/case/_import/views.py [hunk 5]
```diff
@@ -649,21 +651,74 @@ def prepare_response(request, pk):
...
+
+    return render(
+        request=request,
+        template_name="web/domains/case/import/manage/prepare-firearms-oil-response.html",
+        context=context,
+    )
+
+
+def _prepare_sanctions_and_adhoc_response(
+    request: HttpRequest, application: SanctionsAndAdhocApplication, context: dict
```
Comment (oskusalerma) [inline web/domains/case/_import/views.py:687]: same

File: web/domains/case/_import/views.py [hunk 6]
```diff
@@ -649,21 +651,74 @@ def prepare_response(request, pk):
...
+
+    return render(
+        request=request,
+        template_name="web/domains/case/import/manage/prepare-sanctions-response.html",
+        context=context,
+    )
+
+
+def _prepare_derogations_response(
+    request: HttpRequest, application: DerogationsApplication, context: dict
```
Comment (oskusalerma) [inline web/domains/case/_import/views.py:701]: same

File: web/domains/case/_import/views.py [hunk 7]
```diff
@@ -649,21 +651,74 @@ def prepare_response(request, pk):
...
+
+    return render(
+        request=request,
+        template_name="web/domains/case/import/manage/prepare-derogations-response.html",
+        context=context,
+    )
+
+
+def _prepare_wood_quota_response(
+    request: HttpRequest, application: WoodQuotaApplication, context: dict
```
Comment (oskusalerma) [inline web/domains/case/_import/views.py:713]: same

File: web/domains/case/_import/wood/views.py [hunk 8]
```diff
@@ -294,3 +295,35 @@ def submit_wood_quota(request, pk: int) -> HttpResponse:
...
 
         return render(request, "web/domains/case/import/wood/submit.html", context)
+
+
+@login_required
+@permission_required("web.reference_data_access", raise_exception=True)
+def edit_goods(request: HttpRequest, pk: int) -> HttpResponse:
+    application: WoodQuotaApplication = get_object_or_404(WoodQuotaApplication, pk=pk)
+    with transaction.atomic():
+        task: Task = application.get_task(
```
Comment (oskusalerma) [inline web/domains/case/_import/wood/views.py:305]: same

File: web/templates/web/domains/case/import/goods.html [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  {% for row in goods_list %}
+    <tr class="result-row">
+      <td>{{ row.commodity_code }}</td>
+      <td>{{ row.goods_description }}</td>
+      <td>{{ row.quantity_amount }}</td>
+      <td>{{ row.unit }}</td>
+      <td>{{ row.value }}</td>
+      <td>
+        <a class="link-button button icon-pencil" data-input_action="edit"
+          href="{{ icms_url('import:sanctions:edit-goods', kwargs={'application_pk': process.pk, 'goods_pk': row.pk}) }}"></a>
```
Comment (oskusalerma) [inline web/templates/web/domains/case/import/goods.html:24]: filename is generic but this is a specific url...seems iffy?

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/templates/web/domains/case/import/manage/prepare-response.html [hunk 10]
```diff
@@ -49,20 +49,64 @@ <h3>Response Preparation</h3>
...
+          <th>
+            End Date
+          </th>
+          <th>
+            Issue paper licence only?
+          </th>
+          <th>
+            Action
+          </th>
+        </tr>
```
Comment (MattHolmes123) [inline web/templates/web/domains/case/import/manage/prepare-response.html:85]: each <th> element could fit on to one line (Like you do in the next file)

File: web/domains/case/_import/derogations/views.py (hunk 1 above)
Comment (kevinetienne) [reply web/domains/case/_import/derogations/views.py:245]: :+1: it returns `Revealed type is 'web.flow.models.Task'`

File: web/templates/web/domains/case/import/goods.html (hunk 9 above)
Comment (kevinetienne) [reply web/templates/web/domains/case/import/goods.html:24]: First attempt to refactor goods. It needs to be removed.

File: web/domains/case/_import/derogations/views.py (hunk 1 above)
Comment (kevinetienne) [reply web/domains/case/_import/derogations/views.py]: Fixed in the rebase, I've removed the type. Thanks for `reveal_type` I didn't know about it before.

File: web/domains/case/_import/views.py (hunk 4 above)
Comment (kevinetienne) [reply web/domains/case/_import/views.py]: Updated in the rebase, thanks

File: web/templates/web/domains/case/import/manage/prepare-response.html (hunk 10 above)
Comment (kevinetienne) [reply web/templates/web/domains/case/import/manage/prepare-response.html:77]: Updated in rebase

File: web/domains/case/_import/sanctions/views.py [hunk 11]
```diff
@@ -115,42 +117,74 @@ def add_goods(request, pk):
...
+        template_name = "web/domains/case/import/manage/edit-goods-licence.html"
+
+        return _edit_goods(request, application, task, Form, goods_pk, success_url, template_name)
+
+
+def _edit_goods(
+    request: HttpRequest,
+    application: SanctionsAndAdhocApplication,
+    task: Task,
+    Form: forms.ModelForm,
```
Comment (oskusalerma) [inline web/domains/case/_import/sanctions/views.py:159]: - function arguments should be named like_this (i suggest form_class)
- the type of this argument is a class object, so Type[forms.ModelForm]

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/domains/case/_import/sanctions/views.py (hunk 11 above)
Comment (kevinetienne) [reply web/domains/case/_import/sanctions/views.py]: fixed in https://github.com/uktrade/icms/pull/360/commits/175f3809a807e169cb2e36de670a290a2566ccae

**Review by oskusalerma — APPROVED** [review]

**Pull request merged**